### ISSUE TITLE
Multiple users at a time and view-only passwords

### ISSUE THREAD

**Issue by ehfd** [selkies-project/selkies-gstreamer#39]
Apache Guacamole and noVNC provide access to multiple users simultaneously to one screen. They also have functionalities which allow users to set view-only passwords for view-only screen sharing.
It is required to investigate the feasibility of both functionalities for the Selkies project to be used more widely as a stable remote desktop solution.

**Comment by ehfd** [comment]
Relevant to #25 .

**Comment by ehfd** [comment]
Relevant code: https://github.com/selkies-project/selkies-gstreamer/blob/master/src/selkies_gstreamer/signalling_web.py
Should be addressed with #39.

**Comment by ehfd** [comment]
This might be addressed with #48.

**Comment by nums** [comment]
To be able to send stream to others participants, I think the ideal way is to send stream to a WebRTC SFU server.
Neko project is using Pion WebRTC (https://github.com/m1k1o/neko) to achieve that.
Webrtcbin must be able to interface with SFUs (there is a WHIP client that uses webrtcbin https://github.com/meetecho/simple-whip-client)
Once done, the main problem is the management of the quality of the streaming with the recovery of Nack events, PLI etc and this can produce lags or a noticeable drop in the quality of the video.

The mediasoup documentation explains this very well:
https://mediasoup.org/documentation/v3/scalability/#one-to-many-broadcasting

**Comment by ehfd** [comment]
@nums Thank you for the information! We are currently focusing on improving the documentation for more contributor accessibility. We have a goal for this project to potentially attract community contributors from every relevant field, from experienced GStreamer/WebRTC/conferencing software developers to people who simply know a lot of about optimal encoder parameters, and Python/JS learners.

**Comment by ehfd** [comment]
https://github.com/centricular/webrtcsink
May solve this issue as this has its own signalling server.

**Comment by ehfd** [comment]
@nums I guess that you are working on a contribution. We greatly thank your time.

**Comment by nums** [comment]
Yes I'm trying to share the screen in "readonly" mode, but to be able to take the control il will be an other story

**Comment by ehfd** [comment]
> Yes I'm trying to share the screen in "readonly" mode, but to be able to take the control il will be an other story.

That's already something we really need. Thanks a lot!

**Comment by ehfd** [comment]
Relevant: https://github.com/selkies-project/selkies-gstreamer/issues/48#issuecomment-1687298232

**Comment by ehfd** [comment]
WebRTCSink is a wrapper for WebRTCBin, where settings are available in https://gitlab.freedesktop.org/gstreamer/gst-plugins-rs/-/blob/main/net/webrtc/src/webrtcsink/imp.rs

The issue of WHIP/WHEP is that there is no DataChannel for keys, cursors, and other information. Only read-only access would be possible without other transport protocols.

Thus, the immediate property might rather be a multi-user SFU server.

**Need contribution for the SFU server, perhaps utilizing GStreamer's own signalling server.**

https://gstreamer.freedesktop.org/documentation/rswebrtc/index.html#usage

**Comment by PMohanJ** [comment]
The current architecture of selkies-gstreamer consists of two webrtc connections(video and audio) for one single session to a user. So using plugin `webrtcsink` would result in 2*n connections (of course, all are handled by webrtcsink itself) for n number of session?

If so, considering the below note, wouldn't it be resource intense even for as little as 2-3 users?
> It is important to note that at this moment, encoding is not shared between consumers. While this is not on the roadmap at the moment, nothing in the design prevents implementing this optimization.

**Comment by ehfd** [comment]
> So using plugin webrtcsink would result in 2*n connections (of course, all are handled by webrtcsink itself) for n number of session?

This is called an MCU server.

An SFU server distributes the same stream(s) to many clients. I still think webrtcsink's signaling server is still capable of doing this.

https://forasoft.medium.com/p2p-vs-mcu-vs-sfu-what-to-choose-for-a-video-conference-app-fa5cb4a5dd6e
